### ISSUE TITLE
Key config with non US keyboard

### ISSUE THREAD

**Issue by jkn-** [OoliteProject/oolite#116]
Ubuntu 14.04 in an old slow computer, but fast enough for Oolite.
Tested with with Oolite 1.77.1 and 1.80: If I have 
        key_prev_compass_mode                   =";";
        key_next_compass_mode                   =":";
in ~/.Oolite/AddOns/keyconfig.plist, I cannot change 'compass mode'.

If I have:
        key_prev_compass_mode                   ="<";
        key_next_compass_mode                   =">";
I can change it with ';' & ':' keys! Keys '>' & '<' will not do anything.
I have national keyboard with ';' & ':' where most of you have '>' & '<'

So you have somehow hard-coded keyboard type in Oolite!  

Also
    key_mouse_control           = "M";
does not do anything.

**Comment by AnotherCommander** [comment]
I made a small test to verify whether SDL is responsible for what we are seeing or not. The answer is "most likely yes".

I built and run the Example 3-11: Interpreting Key Event Information from this page:
http://www.libsdl.org/release/SDL-1.2.15/docs/html/guideinputkeyboard.html
and run it on a laptop with Italian layout. I can post actual results later, but what I noticed is this:

The keysym.sym codes, which is what we use in Oolite for reading the input are always returning the US keyboard layout codes, even with this very basic example. So, even if we do break the KEYCODE_DOWN EITHER and KEYCODE_UP_BOTH macros, we will still be reading positional information from the keyboard. This is the same regardless of what the language setting is in Windows. I tried both English and Italian language keyboard configurations when doing the test.

The good news is that the keysym.unicode member seems to be getting the correct information, provided that I use the Italian keyboard setting. This means that for example, the backslash of the Italian keyboard, found in the location of the backquote on the US layout is now read as backslash on the unicode keysym member.

The bad news is that as per comment in the example code, SDL_KEYUP does not contain unicode information. Right now I can't think of a way to handle this properly, although it is also true that I did not verify the correctness of that comment myself. Any ideas welcome.

I'll post an example stdout.txt from a test run as soon as I get the opportunity to work again on the Italian layout laptop.

**Comment by cim--** [comment]
That's irritating. It claims - http://wiki.libsdl.org/CategoryKeyboard - that SDLK_* constants are mapped to the current keyboard layout. Experimenting with that C code, it looks like the SDLK_* constants are mapped to the lower-case keyboard layout, but you have to do Shift- and Ctrl- keys yourself, through some form of magic... so on the UK layout I get
`#` -> reported as `#` correctly
`~` -> reported as `~` in unicode, but `#`+`shift` as the keysym+modifier

So what we could do, to get around SDL_KEYUP not containing unicode info, is:
* On SDL_KEYDOWN, if the key has a unicode value (i.e. not an arrow key, modifier key, etc.), pair the unicode value and the keysym. Send an internal Oolite keydown event for the unicode value if available and the keysym if not
* On SDL_KEYUP, if there's a pair with that keysym, send an internal Oolite key up event for its unicode value, and remove the pair, otherwise send the key up event for the keysym.

I think that would work, though it would require a fair bit of rewriting of the existing input routines.

**Comment by hklene** [comment]
Until this will be resolved, could you please add some more comments in the keyconfig.plist file for all shift combinations? E.g.:
```
	key_advanced_nav_array		= "^";		// Shift+6
	key_snapshot			= "*";		// Shift+8
```
On my german keyboard the '^' and '*' are somewhere else, but still Shift+6 and Shift+8 does trigger the function as intended. Now if I change the characters to what is physically printed on those keys ... '&' and '(' the game won't do anything.

It would also be of some help, if the reference manual contained not only the characters to press, but also a diagram of an english keyboard-layout, so people with other layouts can just try to press those buttons, even if they are carry a different label.

Finally I have unused buttons for 'ä', 'ö' and 'ü' ... how would I find out their corresponding scancodes to note them in keyconfig.plist? Writing the unicode character is displayed fine in the internal keyboard overview screen, but they won't act accordingly.

**Comment by phkb** [comment]
I believe this is resolved in #397
